satellite/satellitedb/certdb.go

```diff
@@ -19,82 +22,96 @@ type certDB struct {
 	db *dbx.DB
 }
 
-func (certs *certDB) SavePublicKey(ctx context.Context, nodeID storj.NodeID, publicKey crypto.PublicKey) (err error) {
+func (certs *certDB) SavePublicKey(ctx context.Context, nodeID storj.NodeID, pi *identity.PeerIdentity) (err error) {
 	defer mon.Task()(&ctx)(&err)
-	pubbytes, err := pkcrypto.PublicKeyToPKIX(publicKey)
-	if err != nil {
-		return Error.Wrap(err)
-	}
 
-	_, err = certs.db.Get_CertRecord_By_Publickey(ctx, dbx.CertRecord_Publickey(pubbytes))
+	tx, err := certs.db.Begin()
 	if err != nil {
-		if err == sql.ErrNoRows {
-			return certs.tryAddPublicKey(ctx, nodeID, publicKey)
-		}
 		return Error.Wrap(err)
 	}
 
-	// publickey for the nodeID entry already exists, just return
-	return nil
-}
+	defer func() {
+		if err == nil {
+			err = tx.Commit()
+		} else {
+			err = errs.Combine(err, tx.Rollback())
+		}
+	}()
 
-func (certs *certDB) tryAddPublicKey(ctx context.Context, nodeID storj.NodeID, publicKey crypto.PublicKey) (err error) {
-	defer mon.Task()(&ctx)(&err)
-	// no rows err, so create/insert an entry
-	pubbytes, err := pkcrypto.PublicKeyToPKIX(publicKey)
-	if err != nil {
-		return Error.Wrap(err)
+	if pi == nil {
+		return Error.New("Peer Identity cannot be nil")
 	}
+	chain := encodePeerIdentity(pi)
 
-	// TODO: use upsert here instead of create
-	_, err = certs.db.Create_CertRecord(ctx,
-		dbx.CertRecord_Publickey(pubbytes),
-		dbx.CertRecord_Id(nodeID.Bytes()),
-	)
-	// another goroutine might race to create the cert record, let's ignore that error
-	if pgutil.IsConstraintError(err) || sqliteutil.IsConstraintError(err) {
-		return nil
-	} else if err != nil {
+	var id int64
+	query := `SELECT id FROM certRecords WHERE peer_identity = ?;`
+	err = tx.QueryRow(certs.db.Rebind(query), chain).Scan(&id)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			// create a new entry
+			_, err = tx.Exec(certs.db.Rebind(`INSERT INTO certRecords ( peer_identity, node_id, update_at ) VALUES ( ?, ?, ? );`), chain, nodeID.Bytes(), time.Now())
+			if err != nil {
+				return Error.Wrap(err)
+			}
+			return nil
+		}
 		return Error.Wrap(err)
 	}
 
+	// already public key exists, just return
 	return nil
 }
 
 // GetPublicKey gets the public key of uplink corresponding to uplink id
-func (certs *certDB) GetPublicKey(ctx context.Context, nodeID storj.NodeID) (_ crypto.PublicKey, err error) {
+func (certs *certDB) GetPublicKey(ctx context.Context, nodeID storj.NodeID) (_ *identity.PeerIdentity, err error) {
```

Comment:
See notes in https://github.com/storj/storj/pull/2690#discussion_r310639785. This is sufficient for this, however we should design it such that we can avoid these large requests.

Reply:
I looked at the comment 
"Note, we always load it from the database otherwise it'll add massive amount of traffic to the database.

Similarly there should be one DB call per verification rather than ~100."

am lost on your first comment. What I see is @aleitner is calling GetPublicKey() with parameters and it calls the DB's related function to make a query and get the appropriate info. I am bit lost here. I am at the DB level which makes call directly to DB and gets the information to other higher calling functions.. .. am like an API call. I must certainly be missing something ...

Reply:
Less requests to DB is better for satellite performance. The comment is about making method to get identities for multiple peers with one DB request (sql).  E.g.
`Get(ctx context.Context, nodeIDs []storj.NodeID) (_ []*identity.PeerIdentity, err error) {`

For example then you can ask for all nodes identities from segment piece hashes in one call.

Reply:
Thanks for the clarification... I got it.. updated the code support batch Get of all peeridentites like a transaction.